### ISSUE TITLE
Debian/Ubuntu Error, can't connect to X11 context

### ISSUE THREAD

**Issue by banan314** [federico-terzi/espanso#161]
After using 
```
espanso start
```
and checking 
```
espanso log
```
I get error message:
```
[ERROR] Error, can't connect to X11 context
```
on Linux Mint.

**Comment by banan314** [comment]
I tried `cp /etc/X11/xinit/xinitrc ~`, trying to do what was posted by @a-baghdadi in https://github.com/federico-terzi/espanso/issues/151#issuecomment-568892334_

**Comment by federico-terzi** [comment]
Hey @banan314,
It seems like espanso is executed before `X11` is ready, though I'm not particularly prepared on `xinitrc`. You could try adding a `sleep` before to give X11 some time to initialize, something like: 

```
sleep 3 && espanso start
```

Any reason you are not using the Systemd integration? That should work out of the box on Linux Mint.
Cheers

**Comment by banan314** [comment]
Hi @federico-terzi , unfortunately `sleep 3` didn't work. What do you mean by systemd integration? I simply followed the instructions on your site: https://espanso.org/install/linux/, using `~/.local/bin` as the directory (it's written that it doesn't matter, right?). Why do you say that I don't use systemd integration and how can I check that?

**Comment by federico-terzi** [comment]
Hey @banan314,
Maybe I didn't explained clearly enough. You wanted to add espanso to `xinitrc` to start it automatically at system start right?
In that case, the official approach to achive that ( in systems which support Systemd, and Linux mint should be one of them ), is by registering espanso to Systemd, and this should happen automatically after running `espanso start` the first time, if you give consent. Otherwise, you could try running:

```
espanso register
```

This should register espanso to Systemd and start it automatically when the system boots up.

Let me know if that helped.

Cheers

**Comment by a-baghdadi** [comment]
> I tried `cp /etc/X11/xinit/xinitrc ~`, trying to do what was posted by @a-baghdadi in [#151 (comment)](https://github.com/federico-terzi/espanso/issues/151#issuecomment-568892334)_

Hi, I am using Archlinux and dwm/bspwm , I do not use desktop, I start window manager in terminal by running startx, but in Linuxmint no need to change in xinitrc, do you use wayland instead of X11?

**Comment by banan314** [comment]
> > I tried `cp /etc/X11/xinit/xinitrc ~`, trying to do what was posted by @a-baghdadi in [#151 (comment)](https://github.com/federico-terzi/espanso/issues/151#issuecomment-568892334)_
> 
> Hi, I am using Archlinux and dwm/bspwm , I do not use desktop, I start window manager in terminal by running startx, but in Linuxmint no need to change in xinitrc, do you use wayland instead of X11?

No, I use X11

**Comment by banan314** [comment]
> Hey @banan314,
> Maybe I didn't explained clearly enough. You wanted to add espanso to `xinitrc` to start it automatically at system start right?
> In that case, the official approach to achive that ( in systems which support Systemd, and Linux mint should be one of them ), is by registering espanso to Systemd, and this should happen automatically after running `espanso start` the first time, if you give consent. Otherwise, you could try running:
> 
> ```
> espanso register
> ```
> 
> This should register espanso to Systemd and start it automatically when the system boots up.
> 
> Let me know if that helped.
> 
> Cheers

Hi @federico-terzi 
It does start with the system start, it is not the problem. 
From `/var/log/syslog`:
```
systemd[2681]: Configuration file /home/user/.config/systemd/user/espanso.service is marked world-writable. Please remove world writability permission bits. Proceeding anyway.
systemd[2681]: Reloading.
systemd[2681]: Configuration file /home/user/.config/systemd/user/espanso.service is marked world-writable. Please remove world writability permission bits. Proceeding anyway.
systemd[2681]: Started espanso daemon.
systemd[2681]: espanso.service: Main process exited, code=exited, status=100/n/a
systemd[2681]: espanso.service: Failed with result 'exit-code'.
systemd[2681]: espanso.service: Service hold-off time over, scheduling restart.
systemd[2681]: espanso.service: Scheduled restart job, restart counter is at 1.
systemd[2681]: Stopped espanso daemon.
```
The problem is in the error message. It comes from https://github.com/federico-terzi/espanso/blob/f28874f9b1a2a2bee555112b8b023e08ab25254c/native/liblinuxbridge/bridge.cpp#L87:
```cpp
Display *check_disp = XOpenDisplay(NULL);

    if (!check_disp) {
        return -1;
    }

    XCloseDisplay(check_disp);
    return 1;
```
and is raised here https://github.com/federico-terzi/espanso/blob/f28874f9b1a2a2bee555112b8b023e08ab25254c/src/context/linux.rs#L43
```rust
let x11_available = unsafe {
    check_x11()
};

if x11_available < 0 {
    error!("Error, can't connect to X11 context");
    std::process::exit(100);
}
```

Basically, `XOpenDisplay` doesn't work at all.

**Comment by federico-terzi** [comment]
Hi @banan314,

> From `/var/log/syslog`:
> 
> ```
> systemd[2681]: Configuration file /home/user/.config/systemd/user/espanso.service is marked world-writable. Please remove world writability permission bits. Proceeding anyway.
> systemd[2681]: Reloading.
> systemd[2681]: Configuration file /home/user/.config/systemd/user/espanso.service is marked world-writable. Please remove world writability permission bits. Proceeding anyway.
> systemd[2681]: Started espanso daemon.
> systemd[2681]: espanso.service: Main process exited, code=exited, status=100/n/a
> systemd[2681]: espanso.service: Failed with result 'exit-code'.
> systemd[2681]: espanso.service: Service hold-off time over, scheduling restart.
> systemd[2681]: espanso.service: Scheduled restart job, restart counter is at 1.
> systemd[2681]: Stopped espanso daemon.
> ```

It may sound strange at first, but this behaviour is intended:
* espanso needs to run when the X11 context is initialized
* We tried several flags for Systemd, but none of them was effective in telling the manager to start espanso only when the X11 display was initialized

Because of these reasons, what the espanso service does is the following:
* Attept to start espanso
* If it fails, reschedule a retry in 3 seconds ( up to 3 times ).

What this does in practice is giving enough time to systemd to initialize X11 and restart espanso when it is available. For this reason, the first time espanso.service is spawned it usually fails, but the following times it should work without problems.

Are you sure espanso is not running in your system? ( the easiest way to check is by running `espanso status` ).

Let me know if you have other doubts

**Comment by banan314** [comment]
Hi @federico-terzi ,
Yes I'm sure it doesn't work. `:espanso` doesn't do anything, `espanso status` prints `espanso is not running`. At the moment I'm writing this, it has failed 2071 times.

**Comment by federico-terzi** [comment]
That's pretty strange indeed @banan314. Could you please try running this command in a terminal and post the output here?

```
espanso -v daemon
```

**Comment by banan314** [comment]
> That's pretty strange indeed @banan314. Could you please try running this command in a terminal and post the output here?
> 
> ```
> espanso -v daemon
> ```

Hi there.
That's really strange, but when I used the above command, it started working :open_mouth:. I will post the output for future reference:
```
18:05:51 [ INFO] espanso version 0.4.1
18:05:51 [ INFO] using config path: /home/kamil/.config/espanso
18:05:51 [ INFO] using package path: /home/kamil/.local/share/espanso/packages
18:05:51 [ INFO] starting daemon...
18:05:51 [ INFO] Binded to IPC unix socket: /home/kamil/.local/share/espanso/espanso.sock
18:05:51 [ INFO] espanso is running!
```
Thank you @federico-terzi  for fast and helpful support!

**Comment by federico-terzi** [comment]
Hey @banan314, I'll explain it a bit so it is more clear.
When you launch `espanso start`, what it does is basically invoking `espanso daemon` as a service using Systemd. If you launch `espanso daemon` directly, espanso will work until the terminal is closed. 
So in this case, `espanso daemon` is not really a solution, just a way to find where the problem was. In this case, I still suspect that by the time `espanso start` is invoked, X11 is not initialized, therefore it throws that error.

Cheers
